gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java

```diff
@@ -0,0 +1,828 @@
+    }
+  }
+
+  protected Table createTable(GobblinMetadataChangeEvent gmce, HiveSpec spec) {
+    String schema = gmce.getTableSchema();
+    org.apache.hadoop.hive.metastore.api.Table table = HiveMetaStoreUtils.getTable(spec.getTable());
+    IcebergUtils.IcebergDataAndPartitionSchema schemas = IcebergUtils.getIcebergSchema(schema, table);
+    TableIdentifier tid = TableIdentifier.of(table.getDbName(), table.getTableName());
+    Schema tableSchema = schemas.tableSchema;
+    Preconditions.checkState(tableSchema != null, "Table schema cannot be null when creating a table");
+    PartitionSpec partitionSpec = IcebergUtils.getPartitionSpec(tableSchema, schemas.partitionSchema);
+    Table icebergTable = null;
+    String tableLocation = null;
+    if (useDataLoacationAsTableLocation) {
+      tableLocation = gmce.getDatasetIdentifier().getNativeName() + TABLE_LOCATION_SUFFIX;
+    }
+    try (Timer.Context context = metricContext.timer(CREATE_TABLE_TIME).time()) {
+      icebergTable =
+          catalog.createTable(tid, tableSchema, partitionSpec, tableLocation, IcebergUtils.getTableProperties(table));
+      log.info("Created table {}, schema: {} partition spec: {}", tid, tableSchema, partitionSpec);
+    } catch (AlreadyExistsException e) {
+      log.warn("table {} already exist, there may be some other process try to create table concurrently", tid);
+    }
+    return icebergTable;
+  }
+
+  protected void rewriteFiles(GobblinMetadataChangeEvent gmce, Map<String, Collection<HiveSpec>> newSpecsMap,
+      Map<String, Collection<HiveSpec>> oldSpecsMap, Table table, TableMetadata tableMetadata) throws IOException {
+    PartitionSpec partitionSpec = table.spec();
+    Transaction transaction = tableMetadata.transaction.get();
+    Set<DataFile> newDataFiles = new HashSet<>();
+    getIcebergDataFilesToBeAdded(gmce.getNewFiles(), partitionSpec, newSpecsMap,
+        IcebergUtils.getSchemaIdMap(getSchemaWithOriginId(gmce), table.schema())).stream()
+        .filter(dataFile -> tableMetadata.addedFiles.getIfPresent(dataFile.path()) == null)
+        .forEach(dataFile -> {
+          newDataFiles.add(dataFile);
+          tableMetadata.addedFiles.put(dataFile.path(), "");
+        });
+    Set<DataFile> oldDataFiles = getIcebergDataFilesToBeDeleted(gmce, table, newSpecsMap, oldSpecsMap, partitionSpec);
+    if (oldDataFiles.isEmpty() && !newDataFiles.isEmpty()) {
+      //We randomly check whether one of the new data files already exists in the db to avoid duplication of re-write events
+      DataFile file = newDataFiles.iterator().next();
+      Expression exp = Expressions.startsWith(ICEBERG_FILE_PATH_COLUMN, (String) file.path());
+      if (FindFiles.in(table).withMetadataMatching(exp).collect().iterator().hasNext()) {
+        //This means this re-write event is duplicated with the one we already handled, so directly return
+        return;
+      }
+      //This is the case we register daily partition to additional database
```

Comment:
Not following this line of comment. Isn't this case a very general case when rewrite is called while there's no existing data files in the table ?